pkg/agent/route/route_windows.go

```diff
@@ -174,6 +191,103 @@ func (c *Client) DeleteRoutes(podCIDR *net.IPNet) error {
 	return nil
 }
 
+func (c *Client) AddServiceVirtualIPToGWRoutes() {
+	metric := util.MetricHigh
+	linkIndex := c.nodeConfig.GatewayConfig.LinkIndex
+	gwIP := c.nodeConfig.GatewayConfig.IPv4
+
+	// A virtual IP is used to avoid too many ARP responders in OVS flows.
+	// Route: Virtual IP (169.254.169.252) -> antrea-gw0
+	vRoute := &util.Route{
+		LinkIndex:         linkIndex,
+		DestinationSubnet: &serviceGWVirtualIPNet,
+		GatewayAddress:    gwIP,
+		RouteMetric:       metric,
+	}
+	util.NewNetRoute(vRoute)
+
+	// For ClusterIP Service from host, reply traffic needs a route entry to route packet to vIP via antrea-gw0.
+	// If the NextHop of a route is antrea-gw0, then set it as 0.0.0.0. As a result, on-link/0.0.0.0 is in PersistentStore.
+
+	repRoute := &util.Route{
+		LinkIndex:         linkIndex,
+		DestinationSubnet: &config.ServiceGWHairpinIPv4Net,
+		GatewayAddress:    net.IPv4zero,
+		RouteMetric:       metric,
+	}
+	util.NewNetRoute(repRoute)
+}
+
+func (c *Client) AddClusterIPRoute(svcIP net.IP) error {
+	obj, found := c.hostRoutes.Load(svcIP.String())
+	svcIPNet := util.Get32BitMaskCIDR(svcIP)
+	metric := util.MetricHigh
+
+	// Route: SVC IP -> Virtual IP (169.254.169.252)
+	route := &util.Route{
+		LinkIndex:         c.nodeConfig.GatewayConfig.LinkIndex,
+		DestinationSubnet: &svcIPNet,
+		GatewayAddress:    ServiceGWVirtuallP,
+		RouteMetric:       metric,
+	}
+	if found {
+		existingRoute := obj.(*util.Route)
+		if existingRoute.GatewayAddress.Equal(route.GatewayAddress) && existingRoute.RouteMetric == route.RouteMetric {
+			klog.V(4).InfoS("Route already exists", "Service IP", route.DestinationSubnet.String(),
+				"Gateway", route.GatewayAddress.String(), "RouteMetric", route.RouteMetric)
+			return nil
+		}
+		// Remove the existing route if gateway or metric is not as expected.
+		if err := util.RemoveNetRoute(existingRoute); err != nil {
+			klog.ErrorS(err, "Failed to delete existing route entry",
+				"destination", existingRoute.DestinationSubnet.String(),
+				"gateway", existingRoute.GatewayAddress.String())
+			return err
+		}
+	}
+
+	util.NewNetRoute(route)
+	// Remove the Service route on host by kube-proxy.
+	// TODO: When kube-proxy is not used completely, remove the code below.
+	if err := util.RemoveInterfaceAddress(fmt.Sprintf("%s (HNS Internal NIC)", util.ContainerVNICPrefix), svcIP); err != nil {
+		// We need to clean up routes just added if failure here.
+		util.RemoveNetRoute(route)
+		return err
+	}
+
+	c.hostRoutes.Store(route.DestinationSubnet.String(), route)
+	klog.V(2).InfoS("Route is added", "Service IP", route.DestinationSubnet.String(), "Gateway IP",
+		route.GatewayAddress.String())
+	return nil
+}
+
+// replaceNetRoute adds the expected route by force.
```

Comment:
by force -> and it overrides the existing route with the same destination.

Reply:
Updated.